src/sandstorm/util.h

```diff
@@ -334,6 +335,86 @@ class SubprocessSet {
   kj::Promise<void> waitLoop();
 };
 
+class CapRedirector
+    : public capnp::Capability::Server, public kj::Refcounted {
+  // A capability which forwards all calls to some target. If the target becomes disconnected,
+  // the capability queues new calls until a new target is provided.
+  //
+  // We use this to handle the fact that the front-end is allowed to restart without restarting
+  // all grains. The SandstormCore capability -- provided by the front-end -- will temporarily
+  // become disconnected in these cases. We know the front-end will come back up and reestablish
+  // the connection soon, but there's nothing we can do except wait, and in the meantime we don't
+  // want to spurriously fail calls.
+
+public:
+  CapRedirector(kj::PromiseFulfillerPair<capnp::Capability::Client> paf =
+                kj::newPromiseAndFulfiller<capnp::Capability::Client>())
+      : target(kj::mv(paf.promise)),
+        fulfiller(kj::mv(paf.fulfiller)) {}
+
+  uint setTarget(capnp::Capability::Client newTarget) {
+    ++iteration;
+    target = newTarget;
+
+    // If the previous target was a promise target, fulfill it.
+    fulfiller->fulfill(kj::mv(newTarget));
+
+    return iteration;
+  }
+
+  void setDisconnected(uint oldIteration) {
+    if (iteration == oldIteration) {
+      // Our current client was disconnected.
+      ++iteration;
+      auto paf = kj::newPromiseAndFulfiller<capnp::Capability::Client>();
+      target = kj::mv(paf.promise);
+      fulfiller = kj::mv(paf.fulfiller);
+    }
+  }
+
+  kj::Promise<void> dispatchCall(
+      uint64_t interfaceId, uint16_t methodId,
+      capnp::CallContext<capnp::AnyPointer, capnp::AnyPointer> context) override {
+    capnp::AnyPointer::Reader params = context.getParams();
+    auto req = target.typelessRequest(interfaceId, methodId, params.targetSize());
+    req.set(params);
+    return context.tailCall(kj::mv(req));
+  }
+
+private:
+  uint iteration = 0;
+  capnp::Capability::Client target;
+  kj::Own<kj::PromiseFulfiller<capnp::Capability::Client>> fulfiller;
+};
+
+class TwoPartyServerWithBootstrap: private kj::TaskSet::ErrorHandler {
```

Comment:
It took me a bit to figure out how this was different from TwoPartyServer, because I was thinking "but all servers have a bootstrap". But the difference is that it accepts a client bootstrap.

So maybe call this `TwoPartyServerWithClientBootstrap`, and update the comment to explain this.